README.markdown
# YAML SortKey

YAML Sortkey is an atom package that not only automatically validates your YAML files while you develop, but also gives
you the option to automatically sort the keys on file save.

**Full Demo:** [https://www.youtube.com/watch?v=s-GNDcK0Nq0](https://www.youtube.com/watch?v=s-GNDcK0Nq0)

![](http://i.imgur.com/UhByWdu.gif)



### Installation
* Install via Atom's Package manager by searching for "yaml-sort"

### Usage
###### Valid Commands
* `Yaml Sortkey: Validate`
* `Yaml Sortkey: Sort Keys`

![YAML SortKey Valid Commands](https://www.evernote.com/shard/s714/sh/e0e0ff4a-6efd-4f74-920a-7553838a7956/a74026802aed2fad/res/a066cb82-e19f-426e-8f83-84dfd00fb93f/skitch.png?resizeSmall&width=832 "YAML SortKey")

###### Configuration

* **Autosort Directories:**  Accepts an array of file paths in which all yaml files within those paths will be autosorted before saving.

![YAML SortKey Configuration](https://www.evernote.com/shard/s714/sh/4d59a7d3-bf0a-416c-84af-26a609d1ee24/ef234c00ebcc032d/res/69d92f94-9826-47bf-ad0b-babb3f2cd3c9/skitch.png?resizeSmall&width=832 "YAML SortKey Configuration")


CHANGELOG.md
## 0.5.0 - New fork
* Fix safeLoad and safeDump error

## 0.2.0 - Second Release (Fork)
* Make maxWidth option configurable

## 0.1.0 - First Release
* Every feature added
* Every bug fixed


spec/yaml-sort.js
'use babel';

import YamlSortkey from '../lib/yaml-sort';

// Use the command `window:run-package-specs` (cmd-alt-ctrl-p) to run specs.
//
// To run a specific `it` or `describe` block add an `f` to the front (e.g. `fit`
// or `fdescribe`). Remove the `f` to unfocus the block.

describe('YamlSortkey', () => {
  let workspaceElement, activationPromise;

  beforeEach(() => {
    workspaceElement = atom.views.getView(atom.workspace);
    activationPromise = atom.packages.activatePackage('yaml-sort');
  });

  describe('when the yaml-sort:toggle event is triggered', () => {
    it('hides and shows the modal panel', () => {
      // Before the activation event the view is not on the DOM, and no panel
      // has been created
      expect(workspaceElement.querySelector('.yaml-sort')).not.toExist();

      // This is an activation event, triggering it will cause the package to be
      // activated.
      atom.commands.dispatch(workspaceElement, 'yaml-sort:toggle');

      waitsForPromise(() => {
        return activationPromise;
      });

      runs(() => {
        expect(workspaceElement.querySelector('.yaml-sort')).toExist();

        let yamlSortkeyElement = workspaceElement.querySelector('.yaml-sort');
        expect(yamlSortkeyElement).toExist();

        let yamlSortkeyPanel = atom.workspace.panelForItem(yamlSortkeyElement);
        expect(yamlSortkeyPanel.isVisible()).toBe(true);
        atom.commands.dispatch(workspaceElement, 'yaml-sort:toggle');
        expect(yamlSortkeyPanel.isVisible()).toBe(false);
      });
    });

    it('hides and shows the view', () => {
      // This test shows you an integration test testing at the view level.

      // Attaching the workspaceElement to the DOM is required to allow the
      // `toBeVisible()` matchers to work. Anything testing visibility or focus
      // requires that the workspaceElement is on the DOM. Tests that attach the
      // workspaceElement to the DOM are generally slower than those off DOM.
      jasmine.attachToDOM(workspaceElement);

      expect(workspaceElement.querySelector('.yaml-sort')).not.toExist();

      // This is an activation event, triggering it causes the package to be
      // activated.
      atom.commands.dispatch(workspaceElement, 'yaml-sort:toggle');

      waitsForPromise(() => {
        return activationPromise;
      });

      runs(() => {
        // Now we can test for view visibility
        let yamlSortkeyElement = workspaceElement.querySelector('.yaml-sort');
        expect(yamlSortkeyElement).toBeVisible();
        atom.commands.dispatch(workspaceElement, 'yaml-sort:toggle');
        expect(yamlSortkeyElement).not.toBeVisible();
      });
    });
  });
});


spec/easysort-yaml-spec.js
'use babel';

import YamlSortkey from '../lib/easysort-yaml';

// Use the command `window:run-package-specs` (cmd-alt-ctrl-p) to run specs.
//
// To run a specific `it` or `describe` block add an `f` to the front (e.g. `fit`
// or `fdescribe`). Remove the `f` to unfocus the block.

describe('YamlSortkey', () => {
  let workspaceElement, activationPromise;

  beforeEach(() => {
    workspaceElement = atom.views.getView(atom.workspace);
    activationPromise = atom.packages.activatePackage('easysort-yaml');
  });

  describe('when the easysort-yaml:toggle event is triggered', () => {
    it('hides and shows the modal panel', () => {
      // Before the activation event the view is not on the DOM, and no panel
      // has been created
      expect(workspaceElement.querySelector('.easysort-yaml')).not.toExist();

      // This is an activation event, triggering it will cause the package to be
      // activated.
      atom.commands.dispatch(workspaceElement, 'easysort-yaml:toggle');

      waitsForPromise(() => {
        return activationPromise;
      });

      runs(() => {
        expect(workspaceElement.querySelector('.easysort-yaml')).toExist();

        let yamlSortkeyElement = workspaceElement.querySelector('.easysort-yaml');
        expect(yamlSortkeyElement).toExist();

        let yamlSortkeyPanel = atom.workspace.panelForItem(yamlSortkeyElement);
        expect(yamlSortkeyPanel.isVisible()).toBe(true);
        atom.commands.dispatch(workspaceElement, 'easysort-yaml:toggle');
        expect(yamlSortkeyPanel.isVisible()).toBe(false);
      });
    });

    it('hides and shows the view', () => {
      // This test shows you an integration test testing at the view level.

      // Attaching the workspaceElement to the DOM is required to allow the
      // `toBeVisible()` matchers to work. Anything testing visibility or focus
      // requires that the workspaceElement is on the DOM. Tests that attach the
      // workspaceElement to the DOM are generally slower than those off DOM.
      jasmine.attachToDOM(workspaceElement);

      expect(workspaceElement.querySelector('.easysort-yaml')).not.toExist();

      // This is an activation event, triggering it causes the package to be
      // activated.
      atom.commands.dispatch(workspaceElement, 'easysort-yaml:toggle');

      waitsForPromise(() => {
        return activationPromise;
      });

      runs(() => {
        // Now we can test for view visibility
        let yamlSortkeyElement = workspaceElement.querySelector('.easysort-yaml');
        expect(yamlSortkeyElement).toBeVisible();
        atom.commands.dispatch(workspaceElement, 'easysort-yaml:toggle');
        expect(yamlSortkeyElement).not.toBeVisible();
      });
    });
  });
});


spec/yaml-sort-view-spec.js
'use babel';

import YamlSortkeyView from '../lib/yaml-sort-view';

describe('YamlSortkeyView', () => {
  it('has one valid test', () => {
    expect('life').toBe('easy');
  });
});


lib/yaml-sort.js
'use babel';

import { CompositeDisposable } from 'atom';

export default {

  subscriptions: null,

  config: {
      "autosort": {
        "title": "Autosort Directories",
        "type": "array",
        "default": [],
        "items": {
          "type": "string"
        }
      },
      "lineMaxLen": {
        "title": "Maximum line width",
        "type": "integer",
        "default": 80,
        "minium": 1,
        "maximum": 1000
      },
      "noCompatibilityMode": {
        "title": "Disable compatibility mode (don't try to be compatible with old YAML versions)",
        "type": "boolean",
        "default": false,
      }
    },

  activate(state) {
    this.subscriptions = new CompositeDisposable();
    this.subscriptions.add(atom.workspace.observeTextEditors((function(_this){
      return function(editor){
        buffer = editor.getBuffer();
        if((editor.getGrammar()).name == "YAML"){
          buffer.onWillSave(function(){
            _this.validate_buffer(false)
            path = buffer.getPath();
            var config_paths = atom.config.get("yaml-sort.autosort");
            for(var i=0; i < config_paths.length; i++){
              if (path.indexOf(config_paths[i]) > -1){
                  _this.sort_keys();
              }
            }
          })
        }
      }
    })(this)));

    this.subscriptions.add(atom.commands.add('atom-workspace', {
      'yaml-sort:sort-keys': () => this.sort_keys()
    }));

    this.subscriptions.add(atom.commands.add('atom-workspace', {
      'yaml-sort:validate': () => this.validate_buffer(true)
    }));
  },

  destroy(){
    this.subscriptions.dispose();
  },

  validate_buffer(show_notification){
    editor = atom.workspace.getActiveTextEditor();
    yaml_parser = require('js-yaml')

    try {
      var doc = yaml_parser.load(editor.getText());
      if(show_notification){
        atom.notifications.addSuccess("YAML is valid.")
      }
    } catch (e) {
      atom.notifications.addError("YAML is invalid", {detail: e.message, dismissable: true})
      console.log(e);
    }
  },

  sort_keys() {
    yaml_parser = require('js-yaml')
    editor = atom.workspace.getActiveTextEditor();

    try {
      var doc = yaml_parser.load(editor.getText());
      editor.setText(yaml_parser.dump(doc, {
        sortKeys: true,
        lineWidth: atom.config.get('yaml-sort.lineMaxLen'),
        noCompatMode: atom.config.get('yaml-sort.noCompatibilityMode')
      }));
      atom.notifications.addSuccess("Keys resorted successfully")
    } catch (e) {
      atom.notifications.addError("Keys could not be resorted", {detail: e.message, dismissable: true})
      console.log(e);
    }
  },
};


lib/yaml-sortkey.js
'use babel';

import { CompositeDisposable } from 'atom';

export default {

  subscriptions: null,

  config: {
      "autosort": {
        "title": "Autosort Directories",
        "type": "array",
        "default": [],
        "items": {
          "type": "string"
        }
      },
      "lineMaxLen": {
        "title": "Maximum line width",
        "type": "integer",
        "default": 80,
        "minium": 1,
        "maximum": 1000
      }
    },

  activate(state) {
    this.subscriptions = new CompositeDisposable();
    this.subscriptions.add(atom.workspace.observeTextEditors((function(_this){
      return function(editor){
        buffer = editor.getBuffer();
        if((editor.getGrammar()).name == "YAML"){
          buffer.onWillSave(function(){
            _this.validate_buffer(false)
            path = buffer.getPath();
            var config_paths = atom.config.get("yaml-sortkey.autosort");
            for(var i=0; i < config_paths.length; i++){
              if (path.indexOf(config_paths[i]) > -1){
                  _this.sort_keys();
              }
            }
          })
        }
      }
    })(this)));

    this.subscriptions.add(atom.commands.add('atom-workspace', {
      'yaml-sortkey:sort-keys': () => this.sort_keys()
    }));

    this.subscriptions.add(atom.commands.add('atom-workspace', {
      'yaml-sortkey:validate': () => this.validate_buffer(true)
    }));
  },

  destroy(){
    this.subscriptions.dispose();
  },

  validate_buffer(show_notification){
    editor = atom.workspace.getActiveTextEditor();
    yaml_parser = require('js-yaml')

    try {
      var doc = yaml_parser.safeLoad(editor.getText());
      if(show_notification){
        atom.notifications.addSuccess("YAML is valid.")
      }
    } catch (e) {
      atom.notifications.addError("YAML is invalid", {detail: e.message, dismissable: true})
      console.log(e);
    }
  },

  sort_keys() {
    yaml_parser = require('js-yaml')
    editor = atom.workspace.getActiveTextEditor();

    try {
      var doc = yaml_parser.safeLoad(editor.getText());
      editor.setText(yaml_parser.safeDump(doc, {sortKeys: true, lineWidth: atom.config.get('yaml-sortkey.lineMaxLen')}));
      atom.notifications.addSuccess("Keys resorted successfully")
    } catch (e) {
      atom.notifications.addError("Keys could not be resorted", {detail: e.message, dismissable: true})
      console.log(e);
    }
  },
};


lib/easysort-yaml.js
'use babel';

import { CompositeDisposable } from 'atom';

export default {

  subscriptions: null,

  config: {
      "autosort": {
        "title": "Autosort Directories",
        "type": "array",
        "default": [],
        "items": {
          "type": "string"
        }
      },
      "lineMaxLen": {
        "title": "Maximum line width",
        "type": "integer",
        "default": 80,
        "minium": 1,
        "maximum": 1000
      },
      "noCompatibilityMode": {
        "title": "Disable compatibility mode (don't try to be compatible with old YAML versions)",
        "type": "boolean",
        "default": false,
      }
    },

  activate(state) {
    this.subscriptions = new CompositeDisposable();
    this.subscriptions.add(atom.workspace.observeTextEditors((function(_this){
      return function(editor){
        buffer = editor.getBuffer();
        if((editor.getGrammar()).name == "YAML"){
          buffer.onWillSave(function(){
            _this.validate_buffer(false)
            path = buffer.getPath();
            var config_paths = atom.config.get("easysort-yaml.autosort");
            for(var i=0; i < config_paths.length; i++){
              if (path.indexOf(config_paths[i]) > -1){
                  _this.sort_keys();
              }
            }
          })
        }
      }
    })(this)));

    this.subscriptions.add(atom.commands.add('atom-workspace', {
      'easysort-yaml:sort-keys': () => this.sort_keys()
    }));

    this.subscriptions.add(atom.commands.add('atom-workspace', {
      'easysort-yaml:validate': () => this.validate_buffer(true)
    }));
  },

  destroy(){
    this.subscriptions.dispose();
  },

  validate_buffer(show_notification){
    editor = atom.workspace.getActiveTextEditor();
    yaml_parser = require('js-yaml')

    try {
      var doc = yaml_parser.safeLoad(editor.getText());
      if(show_notification){
        atom.notifications.addSuccess("YAML is valid.")
      }
    } catch (e) {
      atom.notifications.addError("YAML is invalid", {detail: e.message, dismissable: true})
      console.log(e);
    }
  },

  sort_keys() {
    yaml_parser = require('js-yaml')
    editor = atom.workspace.getActiveTextEditor();

    try {
      var doc = yaml_parser.safeLoad(editor.getText());
      editor.setText(yaml_parser.safeDump(doc, {
        sortKeys: true,
        lineWidth: atom.config.get('easysort-yaml.lineMaxLen'),
        noCompatMode: atom.config.get('easysort-yaml.noCompatibilityMode')
      }));
      atom.notifications.addSuccess("Keys resorted successfully")
    } catch (e) {
      atom.notifications.addError("Keys could not be resorted", {detail: e.message, dismissable: true})
      console.log(e);
    }
  },
};
